File: Cplusplus/Easy/twoSum.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<int> twoSum(vector<int>& numbers, int target);

int main(int argc, char const *argv[])
{
    vector<int> nums = {-15,-13,-8,-4,-3,-1};
    vector<int> ans = twoSum(nums, -12);
    for (int index : ans) {
        cout << index << ": " << nums[index] << endl;
    }
    return 0;
}

vector<int> twoSum(vector<int>& numbers, int target) {
    int i = 0, j = numbers.size() - 1;

    while (i < j) {
        if (numbers[i] + numbers[j] < target) {
            ++i;
        } else if (numbers[i] + numbers[j] > target) {
            --j;
        } else {
            break;
        }
    }
    return {i, j};
}

File: Python/twoSum.py
"""
Given an array of integers, return indices of the two numbers such that they add up to a specific target.
You may assume that each input would have exactly one solution, and you may not use the same element twice.
"""
from typing import List, Dict


def twoSum(nums, target):
    map = dict()
    idx1 = 0
    idx2 = 0

    for i in range(len(nums)):
        idx2 = i
        if (target - nums[i]) not in map:
            map[nums[i]] = i
        else:
            idx1 = map[target-nums[i]]
            break

    return [idx1, idx2]


def main():
    nums = [2, 7, 11, 15]
    target = 9

    print(twoSum(nums, target))


if __name__ == '__main__':
    main()
